Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.cpp

```diff
@@ -0,0 +1,437 @@
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/MolStandardize/Fragment.h>
+#include <GraphMol/ChemTransforms/ChemTransforms.h>
+#include <RDGeneral/Exceptions.h>
+
+namespace RDKit {
+namespace ScaffoldNetwork {
+
+ScaffoldNetworkParams::ScaffoldNetworkParams(
+    const std::vector<std::string> &bondBreakersSmarts) {
+  bondBreakersRxns.clear();
+  bondBreakersRxns.reserve(bondBreakersSmarts.size());
+  for (auto sma : bondBreakersSmarts) {
+    std::shared_ptr<ChemicalReaction> rxn(RxnSmartsToChemicalReaction(sma));
+    if (rxn.get() == nullptr) {
+      throw ValueErrorException("could not parse reaction smarts: " + sma);
+    }
+    if (rxn->getNumReactantTemplates() != 1) {
+      throw ValueErrorException(
+          "bond breaker reactions must have exactly one reactant");
+    }
+    if (rxn->getNumProductTemplates() != 2) {
+      throw ValueErrorException(
+          "bond breaker reactions must have exactly two products");
+    }
+    rxn->initReactantMatchers();
+    bondBreakersRxns.emplace_back(rxn);
+  }
+};
+
+namespace detail {
+
+void updateMolProps(RWMol &mol, const ScaffoldNetworkParams &params) {
+  RDUNUSED_PARAM(params);
+  MolOps::sanitizeMol(mol);
+}
+
+ROMol *makeScaffoldGeneric(const ROMol &mol, bool doAtoms, bool doBonds) {
+  RWMol *res = new RWMol(mol);
+  if (doAtoms) {
+    for (auto atom : res->atoms()) {
+      atom->setAtomicNum(0);
+      atom->setNumExplicitHs(0);
+      atom->setNoImplicit(false);
+    }
+  }
+  if (doBonds) {
+    for (auto bond : res->bonds()) {
+      bond->setBondType(Bond::SINGLE);
+      bond->getBeginAtom()->setIsAromatic(false);
+      bond->getEndAtom()->setIsAromatic(false);
+    }
+  }
+  return static_cast<ROMol *>(res);
+}
+ROMol *removeAttachmentPoints(const ROMol &mol,
+                              const ScaffoldNetworkParams &params) {
+  RDUNUSED_PARAM(params);
+  RWMol *res = new RWMol(mol);
+  for (unsigned int i = 1; i <= mol.getNumAtoms(); ++i) {
+    auto atom = res->getAtomWithIdx(mol.getNumAtoms() - i);
+    if (!atom->getAtomicNum() && atom->getDegree() == 1) {
+      // if we're removing a neighbor from an aromatic heteroatom,
+      // don't forget to set the H count on that atom:
+      auto nbri = res->getAtomNeighbors(atom);
+      auto nbr = (*res)[*nbri.first];
+      if (nbr->getIsAromatic() && nbr->getAtomicNum() > 0 &&
+          nbr->getAtomicNum() != 6) {
+        nbr->setNoImplicit(true);
+        nbr->setNumExplicitHs(1);
+      }
+      res->removeAtom(atom);
+    }
+  }
+  return static_cast<ROMol *>(res);
+}
+ROMol *pruneMol(const ROMol &mol, const ScaffoldNetworkParams &params) {
+  RDUNUSED_PARAM(params);
+  ROMol *res = MurckoDecompose(mol);
+  res->updatePropertyCache();
+  MolOps::fastFindRings(*res);
+  return res;
+}
+
+ROMol *flattenMol(const ROMol &mol, const ScaffoldNetworkParams &params) {
+  RWMol *res;
+  if (params.flattenKeepLargest) {
+    MolStandardize::LargestFragmentChooser fragmenter;
```

Comment:
This is such a common function that perhaps it belongs in molops

Reply:
Probably worth doing, but not part of this PR, right?

Reply:
Nope.